src/index.ts

```diff
@@ -1,112 +1,205 @@
+    await Segment.track(EventName.ComponentCopied, EventCategory.ClickEvents, {
+      component_type: componentType,
+    });
+  }
+
+  static async trackWhisperClosed() {
+    await Segment.track(EventName.WhisperClosed, EventCategory.Whispers);
+  }
+
+  static async trackWhisperDisplayed(whisperName: string, whisperUpdated: boolean) {
+    await Segment.page(whisperName, whisperUpdated);
+  }
+
+  // To be removed if considered redundant
+  static async trackWhisperTriggered(whisperName: string, triggeredFrom: string) {
+    await Segment.track(EventName.WhisperTriggered, EventCategory.Whispers, {
+      whisper_name: whisperName,
+      triggered_from: triggeredFrom,
+    });
+  }
+
+  /**
+   * Sends the HTTP request to Segment based on given endpoint and body args
+   *
+   * Should not be called directly, use identify, page, or track instead
+   */
+  private static async postRequest(...[endpoint, body]: SegmentRequest) {
+    if (!Segment.initialized) {
+      throw new Error('Segment not initialized');
     }
-    const properties = {
-      loop_name: Segment.loopName,
-      ...props,
-    };
+    const { loopName, userId, writeKey } = Segment.config;
+
     const request: network.HTTPRequest = {
-      url: `${Segment.BASE_URL}/track`,
+      url: `${Segment.BASE_URL}/${endpoint}`,
       method: 'POST',
       headers: {
-        Authorization: [`Basic ${btoa(`${Segment.writeKey}:`)}`],
+        Authorization: [`Basic ${btoa(`${writeKey}:`)}`],
         'Content-Type': ['application/json'],
       },
       body: JSON.stringify({
-        userId: Segment.userId,
-        event,
+        ...body,
+        userId,
         properties: {
-          ...properties,
-          category,
-          label: properties
-            ? Object.entries(properties)
-                .map(([k, v]) => `[${k}: ${v}]`)
-                .join(' ')
-            : undefined,
+          ...body.properties,
+          loop_name: loopName,
         },
       }),
     };
-    const { statusCode, body } = await network.httpRequest(request);
-    if (statusCode >= 200 && statusCode <= 300) {
-      console.log('Successfully added tracking event');
-    } else {
-      console.error(`Unable to send to segment, due to ${statusCode}: ${body}`);
-    }
+
+    await network.httpRequest(request);
   }
+}
 
-  static async trackComponentClicked(
-    whisperName: string,
-    componentType: whisper.WhisperComponentType,
-    componentText: string
+/**
+ * Set a component's handlers to a wrapped version that will call GA then the original handler
+ */
+export function wrap(component: whisper.Component) {
+  /**
+   * This generic function accepts the existing handler and the desired event function, returns a
+   * function that calls the event function and then calls the existing handler if it exists.
+   *
+   * TODO:  Figure out what the handler type can be without making a giant conditional type of all whisper handlers
```

Comment:
Kinda have a few ideas for this but not sure if they work well in TS let me know if you start needing some inspiration on this lol